Require the use of a remote tracking URI for Databricks project runs

smurching opened the pull request:
Require the use of a remote tracking URI accessible to both the local client & Databricks cluster when launching project runs on Databricks

smurching committed:
Don't set tracking URI & run ID for 'mlflow run' command invoked on Databricks if
the client's tracking URI is a local URI. This prevents the remote run from e.g. trying to log metrics / params against /Users/me/mlruns, which doesn't exist
on Databricks

aarondav commented on mlflow/projects/databricks.py:
```diff
@@ -259,9 +259,12 @@ def run_databricks(uri, entry_point, version, parameters, experiment_id, cluster
+    if not tracking.is_local_uri(tracking_uri):
         env_vars[tracking._TRACKING_URI_ENV_VAR] = tracking.get_tracking_uri()
-        env_vars[tracking._RUN_ID_ENV_VAR] = remote_run.run_info.run_uuid
+        run_id = remote_run.run_info.run_uuid
```
What if `remote_run` is None?

smurching committed:
Add regression tests, refactor databricks logic slightly

smurching commented on mlflow/projects/databricks.py:
```diff
@@ -192,23 +192,6 @@ def _run_shell_command_job(project_uri, command, env_vars, cluster_spec):
     return databricks_run_id
 
 
-def _create_databricks_run(tracking_uri, experiment_id, source_name, source_version,
```
Removed in favor of calling `tracking._create_run` directly

aarondav commented on tests/projects/test_databricks.py:
```diff
@@ -210,3 +192,18 @@ def test_fetch_and_clean_project(tmpdir):
             assert handle.read() == "Hello"
+
+
+def test_get_run_env_vars(tmpdir):
```
Could you add a comment on what this is trying to test? Btw, maybe just write the two asserts out, since this seems a bit indirect for what it's actually doing.

aarondav commented on tests/projects/test_databricks.py:
```diff
@@ -210,3 +192,18 @@ def test_fetch_and_clean_project(tmpdir):
+
+def test_get_run_id(tmpdir):
+    for tracking_uri, run_id, expected in [("http://", "a", "a"), (tmpdir.strpath, "b", None)]:
+        assert databricks._get_run_id(tracking_uri, run_id) == expected
```
Maybe write this one out too

smurching committed:
Raise an exception if user launches a Databricks run against a local tracking URI,
update tests

smurching committed:
Fix lint

smurching committed:
Remove unnecessary checks; we can assume (after validations in run_databricks)
that an ActiveRun was created with an accessible tracking server & that experiment ID, tracking URI should be passed to the Databricks job run

aarondav commented:
LGTM!

smurching commented:
Thanks @aarondav! Merging to master :)
